internal/watchers/watcher.go

```diff
@@ -0,0 +1,206 @@
+	describer          StateDescriber
+	hasStarted         bool
+}
+
+const defaultWait = 4 * time.Second
+
+func NewWatcher(stateTransition StateTransition, describer StateDescriber) *Watcher {
+	return &Watcher{
+		stateTransition: stateTransition,
+		describer:       describer,
+	}
+}
+
+func (watcher *Watcher) Watch() error {
+	watcher.hasStarted = false
+	if watcher.ExponentialBackoff {
+		return watcher.exponentialBackoff()
+	}
+
+	return watcher.linearBackoff()
+}
+
+func (watcher *Watcher) exponentialBackoff() error {
+	backoff := defaultWait
+	for {
+		done, err := watcher.IsDone()
+		if err != nil || done {
+			return err
+		}
+		time.Sleep(backoff)
+		backoff *= 2
+	}
+}
+
+func (watcher *Watcher) linearBackoff() error {
+	for {
+		done, err := watcher.IsDone()
+		if err != nil || done {
+			return err
+		}
+		time.Sleep(defaultWait)
+	}
+}
+
+func (watcher *Watcher) IsDone() (bool, error) {
+	if !watcher.stateTransition.HasStartState() {
+		watcher.hasStarted = true
+	}
+
+	state, err := watcher.describer.GetState()
+
+	if !watcher.hasStarted {
+		if !watcher.stateTransition.IsStartState(state) {
+			return false, &InvalidStateError{State: state}
+		}
+		watcher.hasStarted = true
+
+		return false, nil
+	}
+
+	if err != nil {
+		if watcher.stateTransition.IsRetryableError(err) {
+			return false, nil
+		} else if watcher.stateTransition.IsEndError(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	if watcher.stateTransition.IsEndState(state) {
+		return true, nil
+	} else if watcher.stateTransition.IsRetryableState(state) ||
+		watcher.stateTransition.IsStartState(state) {
+		return false, nil
+	}
+	return false, &InvalidStateError{State: state}
+}
+
+type StateTransition struct {
+	StartState          *string
+	EndState            *string
+	EndErrorCode        *string
+	RetryableStates     []string
+	RetryableErrorCodes []string
+}
+
+func (st *StateTransition) HasStartState() bool {
+	return st.StartState != nil
+}
+
+func (st *StateTransition) IsStartState(state string) bool {
+	return st.StartState != nil && state == *st.StartState
+}
+
+func (st *StateTransition) IsEndState(state string) bool {
+	return st.EndState != nil && state == *st.EndState
+}
+
+func (st *StateTransition) IsEndError(err error) bool {
+	if st.EndErrorCode == nil {
+		return false
+	}
+
+	var atlasErr *atlas.ErrorResponse
+	var atlasv2Err *atlasv2.GenericOpenAPIError
+	var errCode string
+
+	if st.EndErrorCode == nil {
+		return false
+	}
+
+	switch {
+	case errors.As(err, &atlasErr):
+		errCode = atlasErr.ErrorCode
+	case errors.As(err, &atlasv2Err):
+		errCode = *atlasv2Err.Model().ErrorCode
+	default:
+		return false
+	}
+
+	return errCode == *st.EndErrorCode
+}
+
+func (st *StateTransition) IsRetryableError(err error) bool {
+	var atlasErr *atlas.ErrorResponse
+	var atlasv2Err *atlasv2.GenericOpenAPIError
+	var errCode string
+
+	switch {
+	case errors.As(err, &atlasErr):
+		errCode = atlasErr.ErrorCode
+	case errors.As(err, &atlasv2Err):
```

Comment:
[nit] We have dedicated methods for fetching error code: https://github.com/mongodb/atlas-sdk-go/blob/main/docs/doc_2_error_handling.md